### ISSUE TITLE
Re-architecting Cloudist to avoid message loss and improve scalability

### ISSUE THREAD

**Issue by ivanvanderbyl** [ivanvanderbyl/cloudist#1]
## TL;DR

Cloudist is broken, and as it's a crucial component in [TestPilot](http://testpilot.me), we are offering a free (forever) account on TestPilot to anyone who fixes it.

**Note**
References to "Ruby AMQP" mean https://github.com/ruby-amqp/amqp
## The Problem

After using Cloudist in production we noticed it would lock up and messages wouldn't be sent or received. This is due partly to the way Ruby AMQP 0.6.7 works and was incorrectly implemented by me in the first instance. Essentially two things (that we are currently aware of) play a factor in this:
1. We open a new Channel every time we reply and never close them. This is an obvious problem, however solving this using the Async design of Ruby AMQP is tricky. One option would be to use Bunny and send all reply messages synchronously, another might be caching channels/queues instead of creating new ones - this is tricky because it needs to be thread safe and work within EventMachine.
2. We were not setting QoS correctly because Ruby AMQP ignores the prefetch option. We now know that we have to call #prefetch(1) on the channel after it connects, something like this:

``` ruby
    @connection = AMQP.connect(:host => "localhost") do
      puts "Connected to AMQP broker. Running #{AMQP::VERSION} version of the gem..."
    end

    @channel  = AMQP::Channel.new(connection) do |channel, open_ok|
      puts "Channel ##{channel.id} is now open!"
      @channel.prefetch(1, false) do
        puts "basic.qos callback has fired"
      end
    end

    @queue    = channel.queue(queue_name, :auto_delete => true)
    @exchange = channel.direct("")

```
## The Solution

Re-architecting Cloudist to be less of a job queue or more of a high-performance two way messaging DSL which we can build a job queue on top of, optionally moving all messaging code to an adapter pattern so we can swap out AMQP for anything else which fits the bill, such as Beanstalkd or even Redis.

Solving the above problems: (1.) We will need to cache the reply queues instead of opening a new one every time we have a message to send, or design it in such a way that we can use Bunny for publishing and AMQP for consuming.  (2.) Use the above code and update Cloudist to support Ruby AMQP 0.8+

**Why can't TestPilot use Xyz job queue**
TestPilot was built from the ground up to provide realtime build feedback, so much so that it will feel like running your tests locally – on the 30 machines you don't have sitting in your office. We don't believe any job queues that currently exist for Ruby allow us to do this.
## The new architecture

I would like to separate Cloudist into 3 components which can be used together or independently.
1. Messaging, this would handle publishing and consuming of messages in a common format, including setting of headers and encoding decoding of payload.
2. Application/Worker, this would be our current Rails integration for class based listeners and workers, essentially building a layer which sits on top of 1. to use Cloudist as a message queue with all the advantages of being able to send realtime replies of progress, logs, errors and updates as the job is processed. This involves a listener in the Rails app and a worker instance deployed somewhere else completely standalone from the Rails app but using Cloudist it should be supplied with all the information it needs to process the job, and optionally do RPC calls back to the Rails app for more info if required.
3. Adapter pattern. It would be great if we could switch out the use of AMQP for Redis or Beanstalkd if desired.

**But isn't re-architecting Cloudist overkill to solving this problem?** On face value yes, but I think it will help simplify the design of Cloudist so that it will be of more benefit to the Ruby community and more than just another job queue.
## The reward

I will give everyone involved in solving this problem free unlimited access to TestPilot once we launch, because this is the one blocking issue stopping us launching.
## How to get involved.
1. Fork Cloudist
2. Start a feature/bugfix branch e.g. git checkout -b feature/my-awesome-idea or support/this-does-not-work
3. Commit and push until you are happy with your contribution
4. Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.
5. Issue a Pull request.

Add comments below.
